javaparser-core/src/main/java/com/github/javaparser/utils/SourceRoot.java

```diff
@@ -68,16 +73,54 @@ public SourceRoot(Path root) {
         assertNotNull(startPackage);
         Log.info("Parsing package \"%s\"", startPackage);
         final Path path = packageAbsolutePath(root, startPackage);
-        Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
+        class AsynchronousParse extends RecursiveAction {
+
+            private static final long serialVersionUID = 1L;
+            private final Path path;
+
+            public AsynchronousParse(Path path) {
+                this.path = path;
+            }
+
             @Override
-            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
-                if (!attrs.isDirectory() && file.toString().endsWith(".java")) {
-                    Path relative = root.relativize(file.getParent());
-                    tryToParse(relative.toString(), file.getFileName().toString());
+            protected void compute() {
+                final List<AsynchronousParse> walks = new ArrayList<>();
+                try {
+                    Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
+                        @Override
+                        public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs)
+                                throws IOException {
+                            if (!dir.equals(AsynchronousParse.this.path)) {
+                                AsynchronousParse w = new AsynchronousParse(dir);
+                                w.fork();
+                                walks.add(w);
+                                return FileVisitResult.SKIP_SUBTREE;
```

Comment:
This piece of code is duplicated below, any chance of cleaning that up? (only if it improves things of course)